force-app/main/default/classes/DMLOperations.cls

```diff
@@ -159,36 +212,69 @@
 	 * @return The Account that was upserted.
 	 */
 	public static Account upsertAccount(String accountName) {
-		// Query for the Account
-		List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName];
-
+		//create new instance of acct obj
+        Account newAcct = new Account();
+        // Query for the Account
+		List<Account> accounts = [SELECT Id FROM Account WHERE Name = :accountName LIMIT 1];
 		// if the Account exists, update the description
+        if (!accounts.isEmpty()){
+             newAcct = accounts.get(0);
+             newAcct.Description = 'Updated Account';
+             upsert newAcct;
+             return newAcct;  
+        
+        } else {
+            // otherwise, create a new Account and set the description
+            newAcct.Description = 'New Account';
+            newAcct.Name = accountName;
+		    // Upsert the Account
+            upsert newAcct;
+            return newAcct;
+            
+        }
 
-		// otherwise, create a new Account and set the description
-
-		// Upsert the Account
-		return null;
 	}
 
     /**
      * Question 10
 	 * Manage the association between Contacts and Accounts
 	 * Ensure that for each Contact in the provided list, there is a corresponding Account Name that matches the Contact's last name.
-	 * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
+	 * If such an Account does not exist, the method creates it; if it does, the method ensures the Contact is associated with it. 
+     * (Hint: use previous upsertAccount method - this may cause bad practices but for learning purposes it's okay.)
      * Example Contact.LastName = 'Doe' Link to Account.Name = 'Doe' 
      * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
 	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
 		// Loop through each Contact
+        for (Contact contact : contactsToUpsert) {
+            //*store the contacts ln into variable
+            String contactLN = contact.LastName;
+            // Extract the Account Name from Contact's LastName
+            //*search for accounts where name matches the contact ln
+            // Get the account based on the name or create one if it doesn't exist
+            List<Account> acctList = new List<Account>([SELECT Id, Name FROM Account WHERE Name = :contactLN LIMIT 1]);
```

Comment:
DML inside a loop. Try using what we learned about Maps to avoid DML within a loop. Review the same code because it should be very similar to this problem.